incubator/virtualcluster/pkg/controller/virtualcluster/master_provisioner_native.go

```diff
@@ -0,0 +1,339 @@
+	args = append(args, "--initial-cluster", icaVal)
+	etcdBdl.StatefulSet.Spec.Template.Spec.Containers[0].Args = args
+}
+
+// completmentAPIServerTemplate completments the apiserver template of the specified clusterversion
+// based on the virtual cluster setting
+func completmentAPIServerTemplate(vcns string, apiserverBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
+	apiserverBdl.StatefulSet.ObjectMeta.Namespace = vcns
+	apiserverBdl.Service.ObjectMeta.Namespace = vcns
+}
+
+// completmentCtrlMgrTemplate completments the controller manager template of the specified clusterversion
+// based on the virtual cluster setting
+func completmentCtrlMgrTemplate(vcns string, ctrlMgrBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
+	ctrlMgrBdl.StatefulSet.ObjectMeta.Namespace = vcns
+}
+
+// deployComponent deploys master component in namespace vcName based on the given StatefulSet
+// and Service Bundle ssBdl
+func (mpn *MasterProvisionerNative) deployComponent(vc *tenancyv1alpha1.Virtualcluster, ssBdl *tenancyv1alpha1.StatefulSetSvcBundle) error {
+	log.Info("deploying StatefulSet for master component", "component", ssBdl.Name)
+
+	ns := conversion.ToClusterKey(vc)
+
+	switch ssBdl.Name {
+	case "etcd":
+		completmentETCDTemplate(ns, ssBdl)
+	case "apiserver":
+		completmentAPIServerTemplate(ns, ssBdl)
+	case "controller-manager":
+		completmentCtrlMgrTemplate(ns, ssBdl)
+	default:
+		return fmt.Errorf("try to deploy unknwon component: %s", ssBdl.Name)
+	}
+
+	err := mpn.Create(context.TODO(), ssBdl.StatefulSet)
+	if err != nil {
+		return err
+	}
+
+	err = controllerutil.SetControllerReference(vc, ssBdl.StatefulSet, mpn.scheme)
+	if err != nil {
+		return err
+	}
+
+	if ssBdl.Service != nil {
+		log.Info("deploying Service for master component", "component", ssBdl.Name)
+		err = mpn.Create(context.TODO(), ssBdl.Service)
+		if err != nil {
+			return err
+		}
+		err := controllerutil.SetControllerReference(vc, ssBdl.Service, mpn.scheme)
+		if err != nil {
+			return err
+		}
+	}
+
+	// make sure the StatsfulSet is ready
+	// (i.e. Status.ReadyReplicas == Spec.Replicas)
+	timeout := time.After(DeployTimeOut)
+	for {
+		period := time.After(ComponentPollPeriod)
+		select {
+		case <-timeout:
+			return fmt.Errorf("deploy %s timeout", ssBdl.Name)
+		case <-period:
+			sts := &appsv1.StatefulSet{}
+			if err := mpn.Get(context.TODO(), types.NamespacedName{
+				Namespace: ns,
+				Name:      ssBdl.Name},
+				sts); err != nil {
+				return err
+			}
+			if sts.Status.ReadyReplicas == *sts.Spec.Replicas {
+				log.Info("component is ready", "component", ssBdl.Name)
+				return nil
+			}
+		}
+	}
+
+}
+
+// createPKISecrets creates secrets to store crt/key pairs and kubeconfigs
+// for master components of the virtual cluster
+func (mpn *MasterProvisionerNative) createPKISecrets(caGroup *vcpki.ClusterCAGroup, namespace string) error {
+	// create secret for root crt/key pair
+	rootSrt := secret.CrtKeyPairToSecret(secret.RootCASecretName,
```

Comment:
go style guide tell us "don't add line breaks to keep lines short when they are more readable long". Can we Keep this statement one line.
https://github.com/golang/go/wiki/CodeReviewComments#line-length

Reply:
line break is removed